Add session management

albgus opened the pull request:
Re-using the same gssapi security context with every request causes
issues with the authentication towards the IPA server. This
implementation will authenticate once and then use the session endpoint
to continue the authentication.

Any request that is successfully authenticated will set a session cookie
in the response that is valid in the /ipa/session/json endpoint.
HTTPClient automatically handles the cookies within the object.

In IPA 4.4 the correct way of authenticating is to issue a request to
the /ipa/session/login_kerberos endpoint, this is not available in 4.2.0
though.

Commit:
Add session management

Re-using the same gssapi security context with every request causes
issues with the authentication towards the IPA server. This
implementation will authenticate once and then use the session endpoint
to continue the authentication.

Any request that is successfully authenticated will set a session cookie
in the response that is valid in the /ipa/session/json endpoint.
HTTPClient automatically handles the cookies within the object.

In IPA 4.4 the correct way of authenticating is to issue a request to
the /ipa/session/login_kerberos endpoint, this is not available in 4.2.0
though.

m4ce reviewed (changes requested):
Thanks for the contribution!

Only concern I have is that this breaks compatibility with 4.2.0 which is widely used. Could you implement a check that looks at the remote API version and only choose this new authentication method if the version if >= 4.4.0?

albgus commented:
This PR is not using the login_kerberos endpoint, because it isn't compatible with 4.2.0 I opted to just send a "ping" command when authenticating. This should work in both 4.2.0 and 4.4.0.

The documentation from Red Hat does however say that the login_kerberos endpoint is the correct way to login, which is why I noted it in the commit message. I might have been a bit unclear on that in the commit message though.

Commit:
Switch to /ipa/session/login_kerberos for login

Actually, it turns out that the login_kerberos endpoint has been
available troughout IPA 4.x. It just not been documented (from what I
could find) until version 4.4.0. It does work with earlier versions
though.

m4ce commented:
Thanks for this contribution!
